packages/flutter/lib/src/material/input.dart

```diff
@@ -13,11 +13,106 @@ import 'theme.dart';
 
 export 'package:sky_services/editing/editing.mojom.dart' show KeyboardType;
 
+/// A container for grouping together multiple Input widgets.
+class Form extends StatefulWidget {
+  Form({
+    Key key,
+    this.child,
+    this.onSubmitted
+  }) : super(key: key) {
+    assert(child != null);
+  }
+
+  /// Called when the input is accepted anywhere on the form.
+  final VoidCallback onSubmitted;
+
+  /// Root of the widget hierarchy that contains this form.
+  final Widget child;
+
+  @override
+  _FormState createState() => new _FormState();
+}
+
+class _FormState extends State<Form> {
+  @override
+  Widget build(BuildContext context) {
+    return new _FormScope(
+      state: this,
+      child: config.child
+    );
+  }
+}
+
+typedef String FormFieldValidator(InputValue value);
+typedef void FormFieldSetter<T>(T newValue);
+
+/// This contains identifying information for Input fields, required if the
+/// Input is part of a Form.
+class FormField<T> {
+  FormField({
+    this.setter,
+    this.validator
+  });
+
+  /// An optional method to call with the new value when the form field changes.
+  final FormFieldSetter<T> setter;
+
+  /// An optional method that validates an input. Returns an error string to
+  /// display if the input is invalid, or null otherwise.
+  final FormFieldValidator validator;
+}
+
+class _FormScope extends InheritedWidget {
+  _FormScope({
+    Key key,
+    Widget child,
+    this.state
+  }) : super(key: key, child: child);
+
+  final _FormState state;
+
+  // We always notify descendants when we rebuild, since we only do so when a
+  // form field changed.
```

Comment:
I used to have a `dirty` bit on _FormState, which was set when an Input changed, and cleared when we rebuilt. This method would then only return true if the dirty bit is set. Do you think it's worth bringing that back? Or is there a better way to do it?